File: python/triangle/triangle.py
"""triangle"""
from itertools import combinations, starmap


def _two_by_two(sides):
    return starmap(
        lambda x, y: (sides[x], sides[y], sides[3 - x - y]), combinations(range(3), 2)
    )


def _is_triangle(sides):
    return all(s > 0 for s in sides) and all(
        i + j > k for i, j, k in _two_by_two(sides)
    )


def equilateral(sides):
    """Is equilateral?"""
    return sides[0] > 0 and set(sides) == {sides[0]}


def isosceles(sides):
    """Is isosceles?"""
    return _is_triangle(sides) and any(i == j for i, j, k in _two_by_two(sides))


def scalene(sides):
    """Is scalene?"""
    return _is_triangle(sides) and not isosceles(sides)

File: javascript/triangle/triangle.js
export class Triangle {
    constructor(a, b, c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }

    kind() {
        if ([this.a <= 0
             , this.b <= 0
             , this.c <= 0
             , this.a + this.b < this.c
             , this.a + this.c < this.b
             , this.b + this.c < this.a
            ].some(id)) {
            return null;
        }
        let equal_sides = [this.a === this.b, this.b === this.c, this.a === this.c].filter(id).length;
        return ["scalene", "isosceles", "equilateral", "equilateral"][equal_sides];
    }
    get isScalene() { return this.kind() === "scalene"; }
    get isIsosceles() { return this.kind() === "isosceles" || this.kind() === "equilateral"; }
    get isEquilateral() { return this.kind() === "equilateral"; }
}

const id = (something) => something;
